### PR TITLE
Add missing platform-specific information on current_dir and set_current_dir

### LINKED ISSUE #93598: Document canonicalization in `std::env::current_dir`
As of now, the stable and nightly documentation on `std::env::current_dir` only states
> Returns the current working directory as a PathBuf.

And proceeds to note a couple of errors the function may fail with.
This does not mention the fact that (as far as I am aware, testing on Linux) the path is also canonicalized, having resolved all symlinks in any segment of the path. I'd imagine not documenting this behavior could lead to confusion and/or bugs, if it already hasn't.

I tried this code within a symlink folder `/example/symlink/here` pointing to `/symlink/target/here`:

```rust
println!("{:#?}",std::env::current_dir().unwrap())
```

I expected to see get the output: `"/example/symlink/here"`
Instead, this was output: `"/symlink/target/here"`

Currently using the nightly compiler.

For obvious reasons, there's no backtrace.

Update: Read a bit through the code and realized this behavior might be libc-dependent
`ldd --version` on my system returns `ldd (GNU libc) 2.33`

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #93598.

### CODE REVIEW

File: library/std/src/env.rs [hunk 1]
```diff
@@ -23,6 +23,11 @@ use crate::sys::os as os_imp;
 
 /// Returns the current working directory as a [`PathBuf`].
 ///
+///# Platform-specific behavior
```
Comment (Amanieu) [inline library/std/src/env.rs:26]:
```suggestion
/// # Platform-specific behavior
```

**Review by Amanieu — COMMENTED** [review]
LGTM apart from minor style nits.

**Comment by GuillaumeGomez** [comment]
@Amanieu Good catch, thanks!

@bors: r=Amanieu rollup

**Comment by AustinScola** [comment]
Thanks for documenting this! I just ran into this, where I didn't expect `std::env::current_dir` to return the canonicalized path (and I actually wanted the non-canonicalized path). This added documentation would probably have helped me figure it out faster.

**Comment by GuillaumeGomez** [comment]
Glad it helped!

**Pull request merged**